src/controllers/user.ts

```diff
@@ -0,0 +1,231 @@
+                yearOfStudy,
+                interests,
+                isAdmin,
+            },
+        });
+        const token = jwt.sign({ id: newUser.id }, JWT_SECRET, { expiresIn: "1h" }); 
+        return res.status(201).json({
+            message: "User registered successfully",
+            user: { id: newUser.id, email: newUser.email },
+            token,
+            success:true
+        });
+    } catch (error) {
+        console.error("Error in signup:", error);
+        return res.status(500).json({ message: "Internal Server Error",success:false });
+    }
+};
+
+ const signin = async (req: Request, res: Response) => {
+    try {
+        const body = req.body;
+        const parseResult = signinBody.safeParse(body);
+        if (!parseResult.success) {
+            return res.status(400).json({
+                message: "Invalid Inputs",
+                success:false
+            });
+        }
+        const { email, password } = parseResult.data;
+        if (!email.endsWith("@iitkgp.ac.in")) {
+            return res.status(400).json({
+                message: "Email must end with @iitkgp.ac.in",
+                success: false,
+            });
+        }
+        const user = await client.user.findUnique({
+            where: { email }
+        });
+
+        if (!user) {
+            return res.status(401).json({ message: "Invalid credentials",success:false});
+        }
+        const isPasswordValid = await bcrypt.compare(password, user.password);
+        if (!isPasswordValid) {
+            return res.status(401).json({ message: "Invalid credentials",success:false });
+        }
+        const token = jwt.sign({ id: user.id }, JWT_SECRET, { expiresIn: "1h" });
+        return res.status(200).json({
+            message: "Signin successful",
+            user: { id: user.id, email: user.email },
+            token,
+            success:true
+        });
+    } catch (error) {
+        console.error("Error during signin:", error);
+        return res.status(500).json({ message: "Internal Server Error" ,success:false});
+    }
+};
+
+const fetchUserDetails = async (req: Request, res: Response) => {
+    const user = req.user; 
+    if (!user) {
+        return res.status(400).json({ message: "User data is not available", success: false });
+    }
+    try {
+        return res.status(200).json({
+             user ,
+             success : true 
+        });
+    } catch (error) {
+        console.error("Error during fetching user details", error);
+        return res.status(500).json({ message: "Internal Server Error", success: false });
+    }
+};
+
+
+const userUpdate = async (req: Request, res: Response) => {
+    const userId=req.user.id;
+    try {
+        const body = req.body;
+        const parseResult = updateUserBody.safeParse(body);
+        if (!parseResult.success) {
+            return res.status(400).json({
+                message: "Invalid inputs",
+                success:false
+            });
```

Comment:
I think it would be better to return the exact error when zod validation fails instead of `Invalid inputs`